[REVIEW] Implement Fuzz tests for cuIO

galipremsagar opened the pull request:
Issue: #6001 

This PR add python side support for fuzz tests in cuIO:

 - [x] Adds support for Random DataFrame generation.
 - [x] Enable logging of fuzz test failures.
 - [x] Add Parquet reader tests
 - [x] Add Parquet writer tests
 - [x] Add CSV reader tests
 - [x] Add CSV writer tests
 - [x] Add JSON reader tests
 - [x] Add JSON writer tests

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

vuule commented on python/cudf/cudf/testing/parquet.py:
```diff
@@ -0,0 +1,199 @@
+        return self._current_params
+
+
+def _generate_rand_meta(obj):
```
this function is very similar between format, should probably be refactored to avoid repetition.

galipremsagar replied:
Moved it to utils.

vuule commented on python/cudf/cudf/testing/parquet.py:
```diff
@@ -0,0 +1,199 @@
+            # We want to operate near the limits of string column
+            # Hence creating a string column of size almost
+            # equal to 2 Billion bytes(sizeof(int))
+            meta["max_string_length"] = 2000000000 / num_rows
```
we don't necessarily want to always have 2B characters, just need to make sure we don't go over. I'd prefer to randomize the string len here and just limit it to the value used here.

galipremsagar replied:
So this param(`max_string_length`) being set in meta will not ensure all the strings generated will be of this length. It's functionality is just the same you have mentioned, it will generate strings of random length less than value in `max_string_length`. 

This value will be fed into `schoice` api call here where this is achieved: https://github.com/rapidsai/cudf/pull/6114/files#diff-6c157a22d6751bd28a543337bd5d49f9R345-R347

vuule replied:
What is the distribution of string lengths? 
My concern is that we have no way to generate a table with a small number of rows and short strings. Having separate variations for rows and lengths resolves this.

galipremsagar replied:
> What is the distribution of string lengths?

[0, 2000000000 / n) `n` being number of rows.



> My concern is that we have no way to generate a table with a small number of rows and short strings. Having separate variations for rows and lengths resolves this.

Okay, makes sense. In that case I'll go ahead with random number generation for `max_string_length` to ensure it is below the limits.

vuule commented:
Let's make sure to document what is missing to close the issue.
What I can see missing:

1. vary readers'/writers' parameters between tests.
2. (average) Run length as a parameter for the generator.
3. List column support (blocked on https://github.com/rapidsai/cudf/pull/6075).

@galipremsagar Anything missing from my list?

galipremsagar commented:
> Let's make sure to document what is missing to close the issue.
> What I can see missing:
> 
>     1. vary readers'/writers' parameters between tests.
> 
>     2. (average) Run length as a parameter for the generator.
> 
>     3. List column support (blocked on #6075).
> 
> 
> @galipremsagar Anything missing from my list?

This list covers it all. 

Addressed 2. Added issue for 1: #6260 and put in a TODO list column support.

kkraus14 commented on python/cudf/cudf/fuzz_tests/fuzz_test_csv.py:
```diff
@@ -0,0 +1,42 @@
+@pythonfuzz(data_handle=CSVWriter)
+def csv_writer_test(gdf):
+    print("csv_writer_test")
+    pd_file_name = "cpu_pdf.csv"
+    gd_file_name = "gpu_pdf.csv"
+
+    pdf = gdf.to_pandas()
+
+    pdf.to_csv(pd_file_name)
+    gdf.to_csv(gd_file_name)
```
Since this is a fuzzer do we actually need to write out to disk or could we use something like a `StringIO` object to keep things nicely in memory?

kkraus14 commented on python/cudf/cudf/fuzz_tests/fuzz_test_json.py:
```diff
@@ -0,0 +1,40 @@
+
+    pdf = gdf.to_pandas()
+
+    pdf.to_json(pd_file_name)
+    gdf.to_json(gd_file_name)
```
I think we specifically want to test `lines=True`

kkraus14 commented on python/cudf/cudf/testing/main.py:
```diff
@@ -0,0 +1,36 @@
+from cudf.testing import fuzzer
```
NVIDIA copyright

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+import copy
```
NVIDIA copyright

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+        self,
+        file_name="temp_csv.csv",
+        dirs=None,
+        max_rows=4096,
```
This seems pretty small, no?

galipremsagar replied:
I had it kept at a smaller default value locally but passed in `max_rows` param value through `pythonfuzz` decorator while running the tests. Any suggestions on what the default has to be?

kkraus14 replied:
I guess if we're always overriding it it isn't a big deal. Another thought could be to default to `None` and then pick a random number in order to add further test cases to the fuzzing tests.

galipremsagar replied:
I haven't committed the overriding changes for all functions so that we(devs) could be able to run these on their workstations.

If we default to `None`, what would the fallback random number be for the size of rows?

kkraus14 replied:
Instead of 100_000, what if we generated a random number between 0 and 100_000? That way we get further test coverage for small corner / edge cases as well?

galipremsagar replied:
Yea, `max_rows` is used for that purpose. The resulting dataframes are of size varying from 0 to `max_rows`.

galipremsagar replied:
That logic is here: https://github.com/rapidsai/cudf/pull/6114/files#diff-baf658dc5cb5bdf7f82b1fee4d5a7ef7R8

kkraus14 commented on python/cudf/cudf/testing/utils.py:
```diff
@@ -0,0 +1,40 @@
+import random
```
NVIDIA copyright

kkraus14 commented on python/cudf/cudf/testing/utils.py:
```diff
@@ -0,0 +1,40 @@
+            # Hence creating a string column of size almost
+            # equal to 2 Billion bytes(sizeof(int))
+            meta["max_string_length"] = random.randrange(
+                0, 2000000000 / num_rows
```
We may want to parameterize this number in case we change the string implementation later

kkraus14 commented on python/cudf/cudf/testing/utils.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def run_test(funcs, args):
+    if len(args) != 2:
+        print("Usage is python file_name.py function_name")
```
Should this throw an exception instead of just printing?

kkraus14 commented on python/cudf/cudf/tests/dataset_generator.py:
```diff
@@ -200,13 +255,173 @@ def generate(
+
+def float_generator(dtype, size):
+    finfo = np.finfo(dtype)
+    return (
+        lambda: np.random.uniform(
+            low=finfo.min / 2, high=finfo.max / 2, size=size,
+        )
+        * 2
```
Why are we multiplying by 2 instead of just generating directly to min / max?

galipremsagar replied:
This is because current NumPy implementation where if we provide limits of `float` we get overflow error.

```python
>>> import numpy as np
>>> finfo = np.finfo("float64")
>>> np.random.uniform(finfo.min, finfo.max, 10)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "mtrand.pyx", line 959, in numpy.random.mtrand.RandomState.uniform
OverflowError: Range exceeds valid bounds
>>> np.random.uniform(finfo.min, finfo.max/2, 10)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "mtrand.pyx", line 959, in numpy.random.mtrand.RandomState.uniform
OverflowError: Range exceeds valid bounds
>>> np.random.uniform(finfo.min/2, finfo.max/2, 10)
array([ 2.80336379e+307, -1.84988527e+307,  6.00097659e+307,
        1.20549184e+307, -2.56658766e+307,  5.69528438e+307,
        4.78719084e+307,  4.34910856e+306,  7.78469138e+307,
       -8.57599676e+306])
>>> np.random.uniform(finfo.min/2, finfo.max/2, 10) * 2
array([ 1.26865884e+307,  7.70631537e+307, -1.46105042e+308,
        1.44618332e+308,  1.59663611e+308,  7.53097671e+307,
       -8.51266949e+307,  1.78258227e+308, -1.05992116e+307,
        1.74548735e+308])
```

kkraus14 commented on python/cudf/cudf/tests/dataset_generator.py:
```diff
@@ -200,13 +255,173 @@ def generate(
+
+
+def datetime_generator(size):
+    return lambda: np.random.randint(low=0, high=2147483647 - 1, size=size,)
```
Should use negatives for datetimes as well since that indicates before epoch, should similarly be able to use `np.iinfo` on `int64` to cover a wider range.

kkraus14 commented on python/cudf/cudf/testing/fuzzer.py:
```diff
@@ -0,0 +1,92 @@
+import datetime
```
NVIDIA copyright

kkraus14 commented on python/cudf/cudf/testing/fuzzer.py:
```diff
@@ -0,0 +1,92 @@
+)
+
+
+try:
+    lru_cache = functools.lru_cache
+except Exception:
+    import functools32
+
+    lru_cache = functools32.lru_cache
```
Why do we need to try / except this? When does `functools.lru_cache` not exist?

galipremsagar replied:
This was a copy-paste. Didn't know why flake8 doesn't complain for this..

kkraus14 commented on python/cudf/cudf/testing/fuzzer.py:
```diff
@@ -0,0 +1,92 @@
+        dirs=None,
+        crash_reports_dir=None,
+        regression=False,
+        max_rows_size=4096,
```
This feels very small

kkraus14 commented on python/cudf/cudf/testing/fuzzer.py:
```diff
@@ -0,0 +1,92 @@
+            try:
+                self._start_time = datetime.datetime.now()
+                self._target(file_name)
+            except Exception as e:
```
We should be more specific about the types of exceptions we handle here. If someone sends a `ctrl + C` to kill the loop we shouldn't catch it here.

galipremsagar replied:
Handled Keyboard interrupts separately and breaking out of the loop. Else, we'd want to log the failure and capture crash params.

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+        self._idx = 0
+        self._current_params = {}
+
+    def _load_params(self, path):
+        with open(path, "r") as f:
+            params = json.load(f)
+        self._inputs.append(params)
```
This method seems shared across all of the Reader / Writer classes, maybe build a parent class to abstract it?

galipremsagar replied:
Done

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+            params = json.load(f)
+        self._inputs.append(params)
+
+    @staticmethod
+    def _rand(n):
+        return random.randrange(1, n)
```
This function too

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+        return random.randrange(1, n)
+
+    def generate_input(self):
+        if self._regression:
+            if self._idx >= len(self._inputs):
+                logging.info(
+                    "Reached the end of all crash.json files to run..Exiting.."
+                )
+                sys.exit(0)
+            param = self._inputs[self._idx]
+            dtypes_meta = param["dtypes_meta"]
+            num_rows = param["num_rows"]
+            num_cols = param["num_columns"]
+            seed = param["seed"]
+            random.seed(seed)
+            self._idx += 1
+            self._current_params = copy.copy(param)
+        else:
+            seed = random.randint(0, 2 ** 32 - 1)
+            random.seed(seed)
+            dtypes_list = list(
+                cudf.utils.dtypes.ALL_TYPES - cudf.utils.dtypes.TIMEDELTA_TYPES
+            )
+            dtypes_meta, num_rows, num_cols = _generate_rand_meta(
+                self, dtypes_list
+            )
+            self._current_params["dtypes_meta"] = dtypes_meta
+            self._current_params["seed"] = seed
+            self._current_params["num_rows"] = num_rows
+            self._current_params["num_columns"] = num_cols
```
It looks like there's a lot of boilerplate here shared between the readers and writers within a file format as well as across formats, maybe we can abstract some of it away?

galipremsagar replied:
Pushed down the common logic into base class. Kept the dtype specific logic in the implemented classes as they are different for some io reader/writers.

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+
+        return df
+
+    @property
+    def current_params(self):
+        return self._current_params
```
Move to parent class

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+            seed = random.randint(0, 2 ** 32 - 1)
+            random.seed(seed)
+            dtypes_list = list(
+                cudf.utils.dtypes.ALL_TYPES - cudf.utils.dtypes.TIMEDELTA_TYPES
```
Is it a TODO to allow TIMEDELTA types through once they're supported in the CSV reader / writer?

galipremsagar replied:
Yes, but now with this issue fixed: https://github.com/rapidsai/cudf/issues/6225 we can remove this.

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,184 @@
+            f"Generating DataFrame with rows: {num_rows} "
+            f"and columns: {num_cols}"
+        )
+        df = dg.rand_dataframe(dtypes_meta, num_rows, seed).to_pandas()
+        df.to_csv(file_name)
```
This will lose any null integer columns for example because we're converting to pandas. Is that okay here?

galipremsagar replied:
I suppose pyarrow doesn't writer for a pyarrow table to csv. In which case the other option is to read into cudf and writ to csv, but that defeats the purpose of using a reference writer.

kkraus14 replied:
We could use Pandas nullable types for integers / strings / booleans. We just wouldn't get coverage of float / timedelta / datetime for now at least.

galipremsagar replied:
Fixed this by having a utility function to convert pyarrow table to pandas nullable types if possible.

kkraus14 commented on python/cudf/cudf/testing/json.py:
```diff
@@ -0,0 +1,180 @@
+            f"Generating DataFrame with rows: {num_rows} "
+            f"and columns: {num_cols}"
+        )
+        df = dg.rand_dataframe(dtypes_meta, num_rows, seed).to_pandas()
+        df.to_json(file_name)
```
Similarly lose nulls here

kkraus14 commented on python/cudf/cudf/testing/parquet.py:
```diff
@@ -0,0 +1,185 @@
+            f"Generating DataFrame with rows: {num_rows} "
+            f"and columns: {num_cols}"
+        )
+        df = dg.rand_dataframe(dtypes_meta, num_rows, seed).to_pandas()
+        df.to_parquet(file_name)
```
We can write parquet directly from PyArrow and retain nulls.

kkraus14 commented on python/cudf/cudf/testing/io.py:
```diff
@@ -0,0 +1,78 @@
+)
+
+
+class IOBase(object):
```
Can we rename this class? `IOBase` is a class in the Python `io` standard library that we don't want to mix up.

kkraus14 commented on python/cudf/cudf/testing/io.py:
```diff
@@ -0,0 +1,78 @@
+import copy
```
copyright

kkraus14 commented on python/cudf/cudf/testing/main.py:
```diff
@@ -0,0 +1,39 @@
+            dirs=kwargs.get("dir", None),
+            crash_reports_dir=kwargs.get("crash_reports_dir", None),
+            regression=kwargs.get("regression", False),
+            max_rows_size=kwargs.get("max_rows_size", 4096),
```
Change value here as well?

kkraus14 commented on python/cudf/cudf/testing/csv.py:
```diff
@@ -0,0 +1,115 @@
+            f"and columns: {num_cols}"
+        )
+        df = dg.rand_dataframe(dtypes_meta, num_rows, seed).to_pandas()
+        df.to_csv(file_name)
```
Could we use a buffer here and in the other Reader class implementations instead of a file as well?

galipremsagar replied:
Should we be having both mechanisms? So that we have coverage for both code-flows while the fuzz tests are run?

kkraus14 replied:
I figured the fuzzing was for finding edge / corner cases in kernels which should be run the same way regardless of whether the input was a file or a buffer. I figured by keeping everything in memory we can avoid expensive IO and get more iterations in.

galipremsagar replied:
Makes sense, changed it :+1:

cwharris reviewed (approved):
~LGTM~

_apparently_ I had a pending review. I figured these comments had been addressed and resolved. I'll make sure they're still relevant.

cwharris commented on python/cudf/cudf/testing/io.py:
```diff
@@ -0,0 +1,77 @@
+        self._max_columns = max_columns
+        self._max_string_length = max_string_length
+
+        for i, path in enumerate(dirs):
```
This will fail if `dirs` is None, which is the default.

galipremsagar replied:
Good catch @cwharris , This was being filtered out in fuzzer so didn't error, added same conditional check here aswell.

cwharris reviewed (changes requested):
Sorry about the lgtm, I thought I had posted and that they'd been resolved. I have a couple of questions. :(

galipremsagar commented on python/cudf/cudf/fuzz_tests/fuzz_test_json.py:
```diff
@@ -0,0 +1,48 @@
+
+    actual = cudf.read_json(pd_file_name)
+    expected = pd.read_json(gd_file_name)
+    assert_eq(actual, expected)
```
This should now be handled.

galipremsagar commented on python/cudf/cudf/fuzz_tests/fuzz_test_csv.py:
```diff
@@ -0,0 +1,50 @@
+
+    actual = cudf.read_csv(pd_file_name)
+    expected = pd.read_csv(gd_file_name)
+    assert_eq(actual, expected)
```
Handling csv and json file contents for now. But parquet binaries look different to compare. So deferring it to be discussed with team and will address it while adding more tests in #6260

galipremsagar commented on python/cudf/cudf/fuzz_tests/fuzz_test_parquet.py:
```diff
@@ -0,0 +1,48 @@
+    pdf.to_parquet(pd_file_name)
+    gdf.to_parquet(gd_file_name)
+
+    actual = cudf.read_parquet(gd_file_name)
+    expected = pd.read_parquet(pd_file_name)
+    assert_eq(actual, expected)
+
+    actual = cudf.read_parquet(pd_file_name)
+    expected = pd.read_parquet(gd_file_name)
+    assert_eq(actual, expected)
```
Parquet binaries look different to compare. So deferring it to be discussed with team and will address it while adding more tests in #6260

kkraus14 commented:
@vuule this should be ready for another look from you as well

vuule commented on python/cudf/cudf/testing/fuzzer.py:
```diff
@@ -0,0 +1,96 @@
+        dirs=None,
+        crash_reports_dir=None,
+        regression=False,
+        max_rows_size=4096,
```
This default is used now?

galipremsagar replied:
The default in `main.py` is the one being used. Changed this one to be uniform too..

galipremsagar commented on python/cudf/cudf/fuzz_tests/fuzz_test_csv.py:
```diff
@@ -0,0 +1,42 @@
+
+    actual = cudf.read_csv(StringIO(pd_buffer))
+    expected = pd.read_csv(StringIO(gd_buffer))
+    assert_eq(actual, expected)
```
Not necessary, removed it.

cwharris reviewed (approved):
lgtm!
